crates/dbs-upcall/kernel/0001-dragonball-introduce-dragonball-subarch.patch

```diff
@@ -0,0 +1,854 @@
++++ b/drivers/misc/Makefile
+@@ -58,3 +58,4 @@ obj-$(CONFIG_ASPEED_LPC_SNOOP)	+= aspeed-lpc-snoop.o
+ obj-$(CONFIG_PCI_ENDPOINT_TEST)	+= pci_endpoint_test.o
+ obj-$(CONFIG_OCXL)		+= ocxl/
+ obj-$(CONFIG_MISC_RTSX)		+= cardreader/
++obj-$(CONFIG_DRAGONBALL_DRIVERS)       += dragonball/
+diff --git a/drivers/misc/dragonball/Kconfig b/drivers/misc/dragonball/Kconfig
+new file mode 100644
+index 000000000000..aa8b84cbf257
+--- /dev/null
++++ b/drivers/misc/dragonball/Kconfig
+@@ -0,0 +1,17 @@
++#
++# Alibaba Dragonball Secure Container Runtime Drivers
++#
++
++menuconfig DRAGONBALL_DRIVERS
++	bool "Alibaba Dragonball Secure Container Runtime Drivers"
++	depends on X86_64 || ARM64
++	default n
++	help
++	  Alibaba Dragonball is a secure container runtime with an embedded micro-vmm
++	  to securely isolate container workloads.
++
++	  Say Y here to get to see options for various misc drivers to support the
++	  Alibaba Dragonball secure container runtime. This option alone does not
++	  add any kernel code.
++
++	  If unsure, say N.
+diff --git a/drivers/misc/dragonball/Makefile b/drivers/misc/dragonball/Makefile
+new file mode 100644
+index 000000000000..1d3ee5729a5d
+--- /dev/null
++++ b/drivers/misc/dragonball/Makefile
+@@ -0,0 +1,6 @@
++# SPDX-License-Identifier: GPL-2.0
++#
++# Makefile for Dragonball misc drivers
++#
++
++obj-$(CONFIG_DRAGONBALL_DRIVERS)	+= dragonball.o
+diff --git a/drivers/misc/dragonball/dragonball.c b/drivers/misc/dragonball/dragonball.c
+new file mode 100644
+index 000000000000..8f1c60c2f2da
+--- /dev/null
++++ b/drivers/misc/dragonball/dragonball.c
+@@ -0,0 +1,265 @@
++// SPDX-License-Identifier: GPL-2.0
++/*
++ * Copyright (C) 2022 Alibaba Cloud. or its affiliates. All Rights Reserved.
++ *
++ * This module introduces dragonball virtual platform device based on a
++ * memory mapped area.
++ *
++ * The guest device may be instantiated in one of two equivalent ways:
++ *
++ * 1. Device Tree node, eg.:
++ *
++ *		dragonball@1e000 {
++ *			compatible = "dragonball";
++ *			reg = <0x1e000 0x100>;
++ *			interrupts = <42>;
++ *		}
++ *
++ * 2. Kernel module (or command line) parameter. Can be used more than once -
++ *    one device will be created for each one. Syntax:
++ *
++ *		[dragonball.]device=<size>@<baseaddr>:<irq>
++ *    where:
++ *		<size>     := size (can use standard suffixes like K, M or G)
++ *		<baseaddr> := physical base address
++ *		<irq>      := interrupt number (as passed to request_irq())
++ *    eg.:
++ *		dragonball.device=0x100@0x100b0000:12
++ *
++ * Based on Virtio MMIO driver.
++ */
++
++#define pr_fmt(fmt) "dragonball: " fmt
++
++#include <asm/setup.h>
++#include <asm/cmdline.h>
++#include <linux/interrupt.h>
++#include <linux/module.h>
++#include <linux/of.h>
++#include <linux/platform_device.h>
++#include <dragonball/dragonball.h>
++
++#define DB_ADD_FEATURE(_feat, _name)					\
++	[DB_FEAT_##_feat] = {						\
++		.name = "dragonball.no"#_name,				\
++		.val = (1 << DB_FEAT_##_feat),				\
```

Comment:
If we have dragonball platform device, do we need to use cmdline to disable featrues?